src/codeu/chat/client/commandline/Chat.java

```diff
@@ -14,310 +14,384 @@
+
+    final Panel panel = new Panel();
+
+    // HELP
+    //
+    // Add a command to print a list of all commands and their description when
+    // the user for "help" while on the root panel.
+    //
+    panel.register("help", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        System.out.println("ROOT MODE");
+        System.out.println("  u-list");
+        System.out.println("    List all users.");
+        System.out.println("  u-add <name>");
+        System.out.println("    Add a new user with the given name.");
+        System.out.println("  u-sign-in <name>");
+        System.out.println("    Sign in as the user with the given name.");
+        System.out.println("  exit");
+        System.out.println("    Exit the program.");
       }
-
-    } else if (token.equals("sign-out")) {
-
-      if (!clientContext.user.hasCurrent()) {
-        System.out.println("ERROR: Not signed in.");
-      } else {
-        signOutUser();
+    });
+
+    // U-LIST (user list)
+    //
+    // Add a command to print all users registered on the server when the user
+    // enters "u-list" while on the root panel.
+    //
+    panel.register("u-list", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        for (final UserContext user : context.allUsers()) {
+          System.out.format(
+              "USER %s (%s)\n",
+              user.user.name,
+              user.user.id);
+        }
       }
-
-    } else if (token.equals("current")) {
-
-      showCurrent();
-
-    } else if (token.equals("u-add")) {
-
-      if (!tokenScanner.hasNext()) {
-        System.out.println("ERROR: Username not supplied.");
-      } else {
-        addUser(tokenScanner.nextLine().trim());
+    });
+
+    // U-ADD (add user)
+    //
+    // Add a command to add and sign-in as a new user when the user enters
+    // "u-add" while on the root panel.
+    //
+    panel.register("u-add", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        final String name = args.hasNext() ? args.nextLine().trim() : "";
+        if (name.length() > 0) {
+          final UserContext user = context.create(name);
+          if (user == null) {
+            System.out.println("ERROR: Failed to create new user");
+          } else {
+            panels.push(createUserPanel(user));
+          }
+        } else {
+          System.out.println("ERROR: Missing <username>");
+        }
       }
-
-    } else if (token.equals("u-list-all")) {
-
-      showAllUsers();
-
-    } else if (token.equals("c-add")) {
-
-      if (!clientContext.user.hasCurrent()) {
-        System.out.println("ERROR: Not signed in.");
-      } else {
-        if (!tokenScanner.hasNext()) {
-          System.out.println("ERROR: Conversation title not supplied.");
+    });
+
+    // U-SIGN-IN (sign in user)
+    //
+    // Add a command to sign-in as a user when the user enters "u-sign-in"
```

Comment:
As per a discussion with Riccardo, I am leaving the verbose comment as what may be easy for use to read, may not be for students. Async operations and callbacks are a more advance concept and explaining when the code will be executed allows those who are not comfortable with the concept to get introduced to it in an easier way.